Adding optional args for every methods

portepa opened the pull request:
I've added args for every function. I'm not sure this is the best way to handle the requests. However the API documentation does not state in all cases if there's query parameter that can be passed. Having it on every function as the last and optional parameter seemed ok to me.

If you have any remarks on this I'll update the code regarding those.

karatechops commented:
This is looking good. This approach is okay for now. I have a couple comments though -

We might want to check the `typeof` in `getUrlParams `, if the user passes in a string instead of an object this method will explode the string and return something crazy. Checking to see if the parameter is an object before mapping may be helpful. This may be a good point to throw an error.

It's not super intuitive to have `args` in various places within the function's arguments, I don't really see a way around this without a breaking change though.

portepa commented:
Good comments,

Wouldn't it be helpful to allow a string parameter or an object? It'll use the string directly as params if provided as string and the object would be parsed in a string otherwise. This is just an idea instead of throwing an error 😄 

Regarding the args, it's indeed weird and we're having a lot of parameters, some other parameters might be added in the future. Would it be nice to have an object with the parameters passed as attributes instead of different parameters of the function?

```
{
  uri: 'uri',
  args: {
     count: '-1',
  }
}
```

for `getDeployments(options)` instead of `getDeployments(uri, args)` for example.

karatechops commented:
You got it @portepa. This structure is what I'm leaning towards as well. It's a breaking change so let's put it on the back burner for v1.0. I'm sure other stuff like this will reveal itself as we use the library more.

Let's push for args as object only, I'm not a huge fan of accepting two data types as it's difficult to document, hard to maintain, and less intuitive for the user. Maybe there's a use case I'm not seeing where string would be beneficial, feel free to provide a scenario.

portepa commented:
You're right, it can only lead to unexpected behaviours. Should the string be encoded? Should it contain the question mark? Object is better and having only that to support will lead in a better codebase for both this library and also the application that will use it.

portepa commented:
I've made an example for getStatus (https://github.com/portepa/hpe-onesphere-js/commit/6383c025eb4b8949dc95c1d5d9abb5b599533295) . Tests are passing correctly, although the getStatus method with arguments correctly formatted has not been tested.

Do you think it's the best way to implement this? If yes, I'll do the same thing for the other methods and commit my changes 😄

karatechops commented:
This looks good but **beware** `typeof` will return object if I passed in an array like `[ 'hi' , 'bar' ]`. Maybe also check if it's not an array with the `isArray` method? `!Array.isArray(args)`

karatechops commented on src/index.js:
```diff
@@ -55,8 +55,12 @@ export default class OneSphere {
-  getStatus() {
-    return fetcher(`${this.host}/rest/status`, { headers: this.headers });
+  getStatus(args) {
+    if (args && typeof args !== 'object') {
```
`&& !Array.isArray(args)`

karatechops commented on src/index.js:
```diff
@@ -77,153 +81,182 @@ export default class OneSphere {
-  getUsers() {
-    return fetcher(`${this.host}/rest/users`, { headers: this.headers });
+  getUsers(args) {
+    const params = args ? getUrlParams(args) : '';
```
Lets DRY this up a bit, we shouldn't be using the same ternary over and over. How about a `handleArgs` method that has the ternary? This way if we every want to handle defining the `params` differently we only need to change one block. Let me know if that doesn't make sense.

karatechops commented:
Hey @portepa, I left some comments in the code. Do you think you can wrap this up soon? I need to do some heavy lifting in the library in preparation for a client visit and want to avoid merge conflicts.

portepa commented:
Yes sorry, I focused on other things at work. Do you mind if I do this tomorrow? You’ll have it for your next day as I’ll be committing this when you sleep :)

karatechops commented:
@portepa sounds great, thank you!

portepa commented:
@karatechops I've made the changes.

I've implemented only one test for getStatus. All tests were passing except for calls with snapshot (APIs calls were successful but answer would be different).

Also I've implemented the conditions for args differently: we needed to throw the error if the object was an array, therefore I did it this way (the one you suggested did not work, or I did not implement it as you would have):
```
if (args && (typeof args !== 'object' || Array.isArray(args)))
```

Is this work ok for you?

karatechops commented on src/index.js:
```diff
@@ -75,128 +75,191 @@ export default class OneSphere {
+    if (args && (typeof args !== 'object' || Array.isArray(args))) {
+      throw new Error('args parameter should be an object');
+    }
+    const params = getUrlParams(args);
```
Each instance of this block should be adjusted to something like:
`const params = handleUrlParams(args);`

`handleUrlParams` would look something like 

```javascript
const handleUrlParams = args => {
  if (!args || typeof args !== 'object' || Array.isArray(args)) {
    return throw new Error('args parameter should be an object');
  }

  return getUrlParams(args);
});
```

If we ever need to change how we handle params we only need to change it in one place.

portepa replied:
Indeed, sorry for that I went too fast. Updating now 😅

karatechops commented:
Hey @portepa when I mentioned we should "DRY this up" in my earlier code comment I refer to DRY as "Don't Repeat Yourself" - if you find yourself copy and pasting the same line of logical comparison code there's probably a better way of doing it. Let me know if this or my code comment doesn't make sense and I can run you through it.

portepa commented:
![image](https://user-images.githubusercontent.com/6834375/37471140-5c8123ba-2869-11e8-9a41-d4948a99d0d7.png)
That should be better... 😅 Hopefully :)
